HiveCatalog and HiveClientPool fixes

rymurr opened the pull request:
* remove option to create pool w/o a defined size
* deprecate last no-arg constructor in HiveCatalog
* move tests to use no-arg constructor

**NOTE:** the `HiveCatalog` copies a new Hadoop Configuration which is a breaking change from how the constructor used to work. See `HiveMetastoreTest` changes for ramifications.

As discussed in https://github.com/apache/iceberg/pull/2180#issuecomment-770062242

rdblue commented on hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:
```diff
@@ -74,9 +74,18 @@
+   * Hive Catalog constructor.
+   *
+   * @deprecated please use the no-arg constructor, setConf and initialize to construct the catalog. Will be removed in
+   * v0.12.0
```
0.12.0 is probably the next release. Can you change this to 0.13.0?

Also, now that 0.11.0 is out, can we remove the methods to remove in 0.12.0?

rymurr replied:
done, also removed the `HadoopCatalog` constructors

rdblue commented on hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalogs.java:
```diff
@@ -34,6 +36,7 @@ private HiveCatalogs() {
     String metastoreUri = conf.get(HiveConf.ConfVars.METASTOREURIS.varname, "");
-    return CATALOG_CACHE.get(metastoreUri, uri -> new HiveCatalog(conf));
+    return CATALOG_CACHE.get(metastoreUri, uri -> (HiveCatalog) CatalogUtil.loadCatalog(HiveCatalog.class.getName(),
+        "hive", ImmutableMap.of(), conf));
```
Minor: I find this line wrapping very awkward because at a glance and from indentation, it looks like these args are passed to `CATALOG_CACHE.get`. But they are actually inside the lambda and passed to `loadCatalog`.

When breaking lines, I prefer to try not to just break wherever a line gets too long and instead break at reasonable points for readability. If possible, I think this should break at `CatalogUtil.loadCatalog` so that the entire method call is on one line. Otherwise, putting all args on one line makes the most sense so that at least the previous line ends with `loadCatalog(` and it is clear that the next line starts the argument list.

Same with the other changes above. Those aren't as bad as this, but I'd still wrap a little differently (not after the first method arg) to increase readability.

rymurr replied:
agreed, fixed.

rdblue commented:
> NOTE: the HiveCatalog copies a new Hadoop Configuration which is a breaking change from how the constructor used to work. See HiveMetastoreTest changes for ramifications.

Can you be more clear? I don't see anything major. What do you mean?

rymurr commented:
> > NOTE: the HiveCatalog copies a new Hadoop Configuration which is a breaking change from how the constructor used to work. See HiveMetastoreTest changes for ramifications.
> 
> Can you be more clear? I don't see anything major. What do you mean?

`setConf` of `HiveCatalog` does `new Configuration(conf)` whereas the constructor we are removing doesn't copy it. To ensure changes to hadoop configs is visible in `HiveCatalog` the user has to change or re-`setConf`. Not a huge issue and probably the correct behaviour as far as we are concerned. But it did require changing a test and may have repercussions for end users.

rymurr commented:
Thanks for the review @rdblue I have updated based on your comments. I have also gone ahead and removed the deprecated constructors but that kicked off a few other changes. Let me know if you prefer that as a separate PR

aokolnychyi commented:
Let me take a look at this now. Sorry I did not get to this one earlier.

aokolnychyi commented:
The change looks correct to me. @rymurr, shall we split this PR into two and address Hadoop and Hive catalogs separately?

rymurr commented:
Thanks for taking a look @aokolnychyi I have removed the hadoop changes and rebased

aokolnychyi commented on flink/src/test/java/org/apache/iceberg/flink/FlinkTestBase.java:
```diff
@@ -57,7 +59,8 @@ public static void startMetastore() {
     metastore.start();
     FlinkTestBase.hiveConf = metastore.hiveConf();
-    FlinkTestBase.catalog = new HiveCatalog(metastore.hiveConf());
+    FlinkTestBase.catalog = (HiveCatalog)
```
I wish we parameterized `loadCatalog`.

aokolnychyi commented on spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java:
```diff
@@ -62,7 +64,8 @@ public static void startMetastoreAndSpark() {
 
-    SparkTestBase.catalog = new HiveCatalog(spark.sessionState().newHadoopConf());
+    SparkTestBase.catalog = (HiveCatalog)
+        CatalogUtil.loadCatalog(HiveCatalog.class.getName(), "hive", ImmutableMap.of(), hiveConf);
```
nit: `hiveConf` is not technically the same as `sessionState().newHadoopConf()` but it does not matter in tests.

aokolnychyi reviewed (approved):
LGTM. Thanks for the work, @rymurr!

I wonder whether we should parameterize `loadCatalog` to avoid extra casts in calls.
For example, like this:

```
public static <T extends Catalog> T loadCatalog(..)
```

aokolnychyi commented on hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:
```diff
@@ -74,78 +73,22 @@
-    initialize(name, props);
+  public HiveCatalog(Configuration conf) {
+    this.name = "hive";
+    int clientPoolSize = conf.getInt(CatalogProperties.CLIENT_POOL_SIZE, CatalogProperties.CLIENT_POOL_SIZE_DEFAULT);
```
@rymurr, shall we use `setConf` and `initialize` like you in `HadoopCatalog` PR here too?
